packages/ui/src/hooks/usePaymentTokensv2.ts

```diff
@@ -0,0 +1,418 @@
+
+  const allPaymentTokens = useMemo(() => {
+    let paymentTokens = chain?.paymentTokens
+
+    if (includeListingCurrency) {
+      const listingCurrencyAlreadyExists = paymentTokens?.some(
+        (token) =>
+          token?.address?.toLowerCase() ===
+            listingCurrency?.contract?.toLowerCase() &&
+          token.chainId === listingCurrencyChainId
+      )
+      if (!listingCurrencyAlreadyExists && listingCurrency) {
+        paymentTokens?.push({
+          ...listingCurrency,
+          decimals: listingCurrency.decimals || 18,
+          address: listingCurrency.contract as Address,
+          chainId: listingCurrencyChainId || 1,
+        })
+      }
+    }
+
+    if (crossChainDisabled) {
+      paymentTokens = paymentTokens?.filter(
+        (token) => token.chainId === chain?.id
+      )
+    }
+
+    if (nativeOnly) {
+      paymentTokens = paymentTokens?.filter(
+        (token) => token.address === zeroAddress
+      )
+    }
+
+    return paymentTokens
+  }, [
+    chain?.paymentTokens,
+    crossChainDisabled,
+    nativeOnly,
+    includeListingCurrency,
+    listingCurrency,
+    listingCurrencyChainId,
+  ])
+
+  const nonNativeCurrencies = useMemo(() => {
+    return allPaymentTokens?.filter(
+      (currency) => currency.address !== zeroAddress
+    )
+  }, [
+    allPaymentTokens,
+    chain?.paymentTokens,
+    crossChainDisabled,
+    nativeOnly,
+    includeListingCurrency,
+  ])
+
+  const nativeCurrencies = useMemo(() => {
+    return allPaymentTokens?.filter(
+      (currency) => currency.address === zeroAddress
+    )
+  }, [
+    allPaymentTokens,
+    chain?.paymentTokens,
+    crossChainDisabled,
+    nativeOnly,
+    includeListingCurrency,
+  ])
+
+  const { data: nonNativeBalances } = useContractReads({
+    contracts: open
+      ? nonNativeCurrencies?.map((currency) => ({
+          abi: erc20ABI,
+          address: currency.address as `0x${string}`,
+          chainId: chainId,
+          functionName: 'balanceOf',
+          args: [address],
+        }))
+      : [],
+    enabled: open,
+    allowFailure: false,
+  })
+
+  const { data: nativeBalances } = useSWR(
+    open ? address : undefined,
+    () => fetchNativeBalances(address, nativeCurrencies),
+    {
+      revalidateOnFocus: false,
+    }
+  )
+
+  const crosschainChainIds = useMemo(() => {
+    if (crossChainDisabled) {
+      return []
+    } else {
+      return (
+        allPaymentTokens
+          ?.filter((token) => token?.chainId !== chain?.id)
+          ?.map((token) => token?.chainId) ?? []
+      )
+    }
+  }, [allPaymentTokens, crossChainDisabled])
+
+  const { data: solverCapacityChainIdMap } = useSolverCapacities(
+    open ? crosschainChainIds : [],
+    chain
+  )
+
+  return useMemo(() => {
+    if (!open) {
+      return []
+    }
+
+    const paymentTokens = allPaymentTokens?.reduce(
+      (tokens, token) => {
+        tokens[`${token.address}:${token.chainId}`] = {
```

Comment:
Fixed this on my sweep modal branch, but need to convert token.address to lowercase here as well